### PR TITLE
Switch to `IntDistribution`

### PR DESCRIPTION
**Pull request by nyanhi**
## Motivation
In short, this PR is a part of issue #2941, and IntDistribution version of PR #3166 .

This PR makes the switch from old `IntUniformDistribution` classes to unified `IntDistribution` by instantiating it in suggest APIs and few other minor places in codebase. Additionally, all instances of old distributions in test suite are replaced with new one, and parametrization over deprecated classes is removed.

## Description of the changes
Please kindly note that only `optuna/distributions.py` and `tests/test_distributions.py` have been pushed at this moment. Will update other changes later.

### CODE REVIEW

**Comment by himkt** [comment]
@nyanhi Please change the base branch from `master` to `feat/unify-distribution` because @xadrianzetx found our PRs are dependent on each other and we want to avoid a CI failure in the master branch.

🔗 https://github.com/optuna/optuna/pull/3166#issuecomment-997390926
🔗 https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request

**Comment by nyanhi** [comment]
Apologies for the delay. Understood the discussion and confirmed this PR is also failing at the `test_storage.py`. Let me try changing the base branch.

On a separate note, 2 files `tests/sampler_tests/test_samplers.py` and `tests/test_transform.py` have not completely been switched to the new distribution yet (due to the same reason with https://github.com/optuna/optuna/pull/3166#issuecomment-997390926)

**Comment by nzw0301** [comment]
I'm afraid that the commit history looks strange; it includes unrelated commits that have already been merged into the master branch. Could you fix it?

**Comment by xadrianzetx** [comment]
I think pulling current master to `optuna:feat/unify-distribution` should resolve that problem.

**Comment by himkt** [comment]
Hi @nyanhi, please let me know if I can help to rebase. 👋

**Comment by nyanhi** [comment]
Thank you @himkt, actually I would like to understand what is next step/how I can help.

**Comment by himkt** [comment]
@nyanhi 

(Sorry I did the rebasing on my own for your PR but) I updated the base branch of your PR.
If this branch needs to be rebased again, you can do it on the GitHub by following:

procedure | img
-- | --
(a) select `master` to sync then (b) select `switch-to-intdist` on `Choose base branch` at the top of this page | スクリーンショット 2022-01-15 202508

Could you please fetch and sync your local environment?
Following command should be enough to sync the workspace.

```sh
@ anywhere
> git checkout switch-to-intdist
Switched to branch 'switch-to-intdist'

@ switch-to-intdist
> git branch -M tmp-switch-to-intdist

@ tmp-switch-to-intdist
> git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.

@ master
> git pull upstream master
From github.com:optuna/optuna
...

@ master
> git fetch upstream pull/3181/head:switch-to-intdist
From github.com:optuna/optuna
 * [new ref]           refs/pull/3181/head -> switch-to-intdist

@ master
> git checkout switch-to-intdist
Switched to branch 'switch-to-intdist'
```

**Comment by nyanhi** [comment]
Thank you for the detailed explanation, let me try this!

**Comment by himkt** [comment]
Do you have any intent not to update here? (`UniformDistribution` is also left but I'm wondering if we have to)
https://github.com/nyanhi/optuna/blob/bfda0583f3efcea4b91a6156fc98b18de3011168/tests/test_transform.py#L55

---

And I just find that the documentation should be updated as well. The same applies to the float distribution, so it would be good to separate from this PR.

- https://github.com/nyanhi/optuna/blob/bfda0583f3efcea4b91a6156fc98b18de3011168/optuna/samplers/_tpe/sampler.py#L94-L109
- https://github.com/nyanhi/optuna/blob/bfda0583f3efcea4b91a6156fc98b18de3011168/optuna/samplers/_tpe/multi_objective_sampler.py#L33-L48
- https://github.com/nyanhi/optuna/blob/bfda0583f3efcea4b91a6156fc98b18de3011168/optuna/multi_objective/samplers/_motpe.py#L37-L52

**Comment by himkt** [comment]
Thank you so much @nyanhi! Basically LGTM. Please check my nit comment.

**Comment by HideakiImamura** [comment]
In addition, could you resolve the conflicts with `feat/unify-distribution`?

File: tests/test_distributions.py [hunk 1]
```diff
@@ -13,29 +12,34 @@
...
     "u": distributions.UniformDistribution(low=1.0, high=2.0),
     "l": distributions.LogUniformDistribution(low=0.001, high=100),
     "du": distributions.DiscreteUniformDistribution(low=1.0, high=9.0, q=2.0),
-    "iu": distributions.IntUniformDistribution(low=1, high=9),
-    "iuq": distributions.IntUniformDistribution(low=1, high=9, step=2),
     "c1": distributions.CategoricalDistribution(choices=(2.71, -float("inf"))),
     "c2": distributions.CategoricalDistribution(choices=("Roppongi", "Azabu")),
     "c3": distributions.CategoricalDistribution(choices=["Roppongi", "Azabu"]),
-    "ilu": distributions.IntLogUniformDistribution(low=2, high=12),
-    "iluq": distributions.IntLogUniformDistribution(low=2, high=12, step=2),
```
[the remark this answers is not in the document]
Comment (nyanhi) [reply tests/test_distributions.py:30]: `IntDistribution` class raises value error when `log` = True and `step` != 1. 
Could you kindly advise if this concerns about backward compatibility for those who has been using `IntLogUniformDistribution` with `step` != 1 ?

File: tests/test_distributions.py [hunk 2]
```diff
@@ -269,57 +258,7 @@ def test_float_contains(expected: bool, value: float, step: Optional[float]) ->
     assert f._contains(value) == expected
 
 
-def test_contains() -> None:
```
[the remark this answers is not in the document]
Comment (nyanhi) [reply tests/test_distributions.py:272]: Thank you for the comment, let me explain my intension:

In the `test_contains()`, multiple classes(e.g. IntUniform, LogUniform) were tested together. However I think it's cleaner to test 3 distribution classes in corresponding test functions with pytest parametrize. Hence `test_int_contains()`, `test_float_contains()`, and `test_categorical_contains()` are introduced.

File: tests/test_distributions.py [hunk 3]
```diff
@@ -441,64 +345,9 @@ def test_float_single(
     assert distribution.single() == expected
 
 
-def test_single() -> None:
```
[the remark this answers is not in the document]
Comment (nyanhi) [reply tests/test_distributions.py:444]: Similar with `test_contains()`, `test_single()` is also replaced by class-wise test functions `test_int_single()`, `test_float_single()`, and `test_categorical_single()`. 

The tests in `test_empty_distribution` are included into `test_int_init_error()` and `test_float_init_error()`. I can add `test_categorical_init_error()`. Please kindly let me know  if it sounds fine with you or not.

File: tests/test_distributions.py [hunk 4]
```diff
@@ -381,26 +305,6 @@ def test_empty_range_contains() -> None:
...
-
-    ilu = distributions.IntLogUniformDistribution(low=1, high=1)
-    assert not ilu._contains(0)
-    assert ilu._contains(1)
-    assert not ilu._contains(2)
-
-    iluq = distributions.IntLogUniformDistribution(low=1, high=1, step=2)
-    assert not iluq._contains(0)
-    assert iluq._contains(1)
-    assert not iluq._contains(2)
```
[the remark this answers is not in the document]
Comment (nyanhi) [reply tests/test_distributions.py:402]: As `step` must be 1 when `log` = True, I think tests for `iluq` are not included.

Given `IntDistribution._contains()` doesn't depend on `log` argument, the test case equivalent for `ilu` can be omitted in my opinion. Please let me know if we should still add it.

File: tests/test_distributions.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (himkt) [reply tests/test_distributions.py:30]: Thank you for explaining. As @nyanhi said, now `IntDistribution` itself validates `low` and `high` and we cannot create a distribution with log and step larger than 1.

I notice it conflicts with `IntLogUniformDistribution` whose `step` is scheduled to be removed in v4.0.0.
But this schedule is not mandatory and we may change `IntLogUnifomDistribution` in v3.0.0 as well.

https://github.com/optuna/optuna/blob/5ec93be4e7b9a307f9ec5636dfaae48cf8fd4c64/optuna/distributions.py#L469-L477

File: tests/test_distributions.py [hunk 5]
```diff
@@ -529,10 +378,6 @@ def test_eq_ne_hash() -> None:
     d1 = distributions.UniformDistribution(low=1, high=3)
     assert d0 != d1
 
-    # Different distribution classes.
-    d2 = distributions.IntUniformDistribution(low=1, high=2)
-    assert d0 != d2
-
```
[the remark this answers is not in the document]
Comment (nyanhi) [reply tests/test_distributions.py:535]: I think so too, shall we change di0 and di1 as well when resolving conflicts?

**Comment by nyanhi** [comment]
@himkt 
Thank you for pointing out, let me update the line you mentioned. Just for my note, following points need to be worked in follow up PRs:

- [ ] Replace old distributions in pytest parametrize for `tests/sampler_tests/test_samplers.py` and `tests/test_transform.py`
- [ ] Updates documentations below
https://github.com/nyanhi/optuna/blob/bfda0583f3efcea4b91a6156fc98b18de3011168/optuna/samplers/_tpe/sampler.py#L94-L109
https://github.com/nyanhi/optuna/blob/bfda0583f3efcea4b91a6156fc98b18de3011168/optuna/samplers/_tpe/multi_objective_sampler.py#L33-L48
https://github.com/nyanhi/optuna/blob/bfda0583f3efcea4b91a6156fc98b18de3011168/optuna/multi_objective/samplers/_motpe.py#L37-L52

**Comment by nyanhi** [comment]
@HideakiImamura

1. Thank you so much for the comments on `test_distributions.py`, please find my responses and kindly let me know if it sounds fine with you or not. If it's ok, I will proceed to resolve the conflicts. Of course, happy to hear your thought!

2. For the grep results, there are 2 major reasons you still see old distributions. 
   (a) `tests/sampler_tests/test_samplers.py` and `tests/test_transform.py` will be updated in a follow up PR.
   (b) isinstance(d, `an old_distributions`) clause is left as is for now, and perhaps we need to discuss when we should change or delete those. If I recall chat with himkt the other day correctly, old distributions can be still instantiated through CLI (sorry, I'm not 100% sure on this point...).

File: tests/test_distributions.py [hunk 6]
```diff
@@ -587,18 +432,6 @@ def test_discrete_uniform_distribution_asdict() -> None:
     assert EXAMPLE_DISTRIBUTIONS["du"]._asdict() == {"low": 1.0, "high": 9.0, "q": 2.0}
```
[the remark this answers is not in the document]
Comment (nyanhi) [reply tests/test_distributions.py:433]: Think it's linking to the discussion above, kindly take a look comments from himkt and myself.

**Comment by himkt** [comment]
> old distributions can be still instantiated through CLI (sorry, I'm not 100% sure on this point...)

It's true. But it turned out that we can remove the old distribution from codebase by making them be wrappers of {Float,Int}Distribution. I didn't notice it so I thought we should leave old distribution occurrences in codebase.

I opened another PR to **`master`** branch #3244. After this PR being merged, this PR can be simplified to remove if-then rules for old distributions from codebase, I think. I'm sorry for confusing you but it's just the thing behind the comment by @HideakiImamura. I'd like to know if it makes sense and if there is something I misunderstand.

**Comment by himkt** [comment]
Now #3244 is merged into the master branch. We are able to remove branchings of old distributions!

**Comment by nyanhi** [comment]
Now old distributions (IntUniform and IntLogUniform) appear only on its definition and below lines in `optuna.distributions.py` which we need to keep. Follow ups https://github.com/optuna/optuna/pull/3181#issuecomment-1018554169 are actually included into this PR. Thank you for all the supports as always, please let me know if any question.

https://github.com/optuna/optuna/pull/3181/files#diff-2ff77057715ff5d7b689aa1ba37562d8a47052b826e0267493bfdd13ac7206e3L499-R508

**Comment by HideakiImamura** [comment]
Thanks for the update! This might be outside of this PR, but the CI is failing due to the following error. Do you think we need to fix the error before this PR or in this PR? Cc: @himkt 

```
dist_old = UniformDistribution(high=5.0, low=-5.0)
dist_new = FloatDistribution(high=5.0, log=False, low=-5.0, step=None)

    def check_distribution_compatibility(
        dist_old: BaseDistribution, dist_new: BaseDistribution
    ) -> None:
        """A function to check compatibility of two distributions.
    
        Note that this method is not supposed to be called by library users.
    
        Args:
            dist_old: A distribution previously recorded in storage.
            dist_new: A distribution newly added to storage.
    
        Raises:
            ValueError:
                If different distribution kinds are set to ``dist_old`` and ``dist_new``,
                or ``dist_old.choices`` doesn't match ``dist_new.choices``
                for :class:`~optuna.distributions.CategoricalDistribution`.
        """
    
        if dist_old.__class__ != dist_new.__class__:
>           raise ValueError("Cannot set different distribution kind to the same parameter name.")
E           ValueError: Cannot set different distribution kind to the same parameter name.

optuna/distributions.py:591: ValueError
```

**Comment by nyanhi** [comment]
@HideakiImamura Thank you, I think the failure of storage test is same reason with https://github.com/optuna/optuna/pull/3166#issuecomment-1009759032.

**Comment by himkt** [comment]
Yes, as @nyanhi said we will tolerate CI failure for storage upgrade since it would not be fixed after merging #3113. These PRs depend on each other so we first merge refactoring PRs into feature branch and then add migration script to it.

File: optuna/distributions.py [hunk 7]
```diff
@@ -548,16 +548,9 @@ def json_to_distribution(json_str: str) -> BaseDistribution:
...
-                    else:
-                        return IntLogUniformDistribution(low, high)
-                else:
-                    if step is not None:
-                        return IntUniformDistribution(low, high, step)
-                    else:
-                        return IntUniformDistribution(low, high)
+                if step is None:
+                    step = 1
+                return IntDistribution(low=low, high=high, log=log, step=step)
```
Comment (himkt) [inline optuna/distributions.py:553]: [note] `json_to_distribution` creates `IntDistribution` for abbreviated format after this PR.

File: optuna/integration/cma.py [hunk 8]
```diff
@@ -277,24 +269,14 @@ def _initialize_sigma0(search_space: Dict[str, BaseDistribution]) -> float:
 
         sigma0s = []
         for name, distribution in search_space.items():
-            if isinstance(distribution, UniformDistribution):
```
Comment (himkt) [inline optuna/integration/cma.py:280]: I think this change (for `FloatDistribution`) should not be included in this PR. Could you please extract changes for `IntDistribution`?

**Review by himkt — COMMENTED** [review]
Thank you so much @nyanhi for revising! Looks basically solid.
BTW, c6663a1 18c7e3e seemingly contains several changes for `FloatDistribution`.
I think this PR should focus on cleaning old distribution for `IntDistribution`, is it possible to extract changes for `IntDistribution` from them?

File: optuna/integration/cma.py (hunk 8 above)
Comment (nyanhi) [reply optuna/integration/cma.py:280]: Thank you for pointing out, sorry while resolving conflicts I went ahead for some changes involves FloatDistribution but will revert it. Some parts of change may conflict with works on FloatDistribution side (e.g. below numbered parametrize), but I understood that we can resolve it later. Thank you for the guidance, will work on it by tomorrow.

https://github.com/optuna/optuna/commit/c6663a1f4ddf11039eee8743bac2924598a660a9#diff-8b85ccf8a3534b2aa878bce8ece8ce930af410ab178ebc6d80031badbabb5052L842-R818

Comment (himkt) [reply optuna/integration/cma.py:280]: Thank you for the quick response and your cooperation @nyanhi. ❤️
This refactoring is complicated and I force you to revise for resolving them, I really appreciate your contribution.

The same refactoring on `FloatDistribution` side is going on by @xadrianzetx.
We will merge this PR (#3181) before #3275, so we will ask @xadrianzetx to resolve conflict if it is needed.
In this PR, please focus on cleaning up `IntDistribution` and its subclasses. :pray:

**Review by himkt — COMMENTED** [review]
Almost LGTM. Please take a look at comments.

File: tests/test_distributions.py [hunk 9]
```diff
@@ -352,64 +298,9 @@ def test_float_single(
...
-        distributions.FloatDistribution(low=-30, high=-40, step=3)
-
-    with pytest.raises(ValueError):
-        distributions.IntUniformDistribution(low=123, high=100)
-
-    with pytest.raises(ValueError):
-        distributions.IntUniformDistribution(low=123, high=100, step=2)
-
-    with pytest.raises(ValueError):
-        distributions.CategoricalDistribution(choices=())
```
[the remark this answers is not in the document]
Comment (nyanhi) [reply tests/test_distributions.py:409]: Thank you very much for pointing out, I recall the discussion now. Apologies.

File: optuna/integration/skopt.py [hunk 10]
```diff
@@ -251,13 +251,9 @@ def __init__(
                 # Convert the upper bound from exclusive (optuna) to inclusive (skopt).
                 high = np.nextafter(distribution.high, float("-inf"))
                 dimension = space.Real(distribution.low, high, prior="log-uniform")
-            elif isinstance(distribution, distributions.IntUniformDistribution):
```
[the remark this answers is not in the document]
Comment (nyanhi) [reply optuna/integration/skopt.py:254]: Thank you for pointing out, makes sense!

File: tests/test_transform.py [hunk 11]
```diff
@@ -84,15 +79,11 @@ def test_search_space_transform_encoding() -> None:
...
-        (3, IntUniformDistribution(0, 3)),
-        (1, IntLogUniformDistribution(1, 10)),
-        (10, IntLogUniformDistribution(1, 10)),
-        (2, IntUniformDistribution(0, 10, step=2)),
         (0, IntDistribution(0, 3)),
+        (3, IntDistribution(0, 3)),
         (1, IntDistribution(1, 10, log=True)),
+        (10, IntDistribution(1, 10, log=True)),
         (2, IntDistribution(0, 10, step=2)),
-        (10, IntUniformDistribution(0, 10, step=2)),
```
[the remark this answers is not in the document]
Comment (nyanhi) [reply tests/test_transform.py:95]: Sure, added the case.

**Comment by nyanhi** [comment]
Thank you so much for swift review and comments, I agree with all of them. 
I noticed mypy error, and made fix in the latest commit [8212d9f](https://github.com/optuna/optuna/pull/3181/commits/8212d9f38f602c5d7cb61c08524f3830d0e8b94e). Please kindly let me know if any concern on this point.

**Review by HideakiImamura — APPROVED** [review]
Thanks for the long running PR! Looks perfect!

**Review by himkt — APPROVED** [review]
Thank you so much for the patient contribution!
Finally LGTM!
